sarif/main.ts

```diff
@@ -0,0 +1,167 @@
+
+type Dependency = {
+    purl: string;
+    registry: string;
+    name: string;
+    version: string;
+    issues: Issue[];
+};
+
+type Issue = {
+    title: string;
+    tag: string;
+    domain: string;
+    severity: string;
+    description: string;
+};
+
+type InputJson = {
+    dependencies: Dependency[];
+};
+
+type Project = {
+    name: string;
+    id: string;
+    updated_at: string;
+    created_at: string;
+    ecosystems: string[];
+    group_name: string;
+};
+
+const accessToken = await PhylumApi.getAccessToken();
+
+/**
+ * Given a Phylum JSON response, convert it into a valid SARIF JSON.
+ */
+const convertToSarif = (input: InputJson) => {
+    const sarif: any = {
+        "$schema": "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json",
+        "version": "2.1.0",
+        "runs": [
+            {
+                "tool": {
+                    "driver": {
+                        "name": "Phylum",
+                        "version": "1.0.0",
+                        "informationUri": "https://phylum.io",
+                        "rules": [] // This will be populated later
+                    }
+                },
+                "results": [] // This will be populated later
+            }
+        ]
+    };
+
+    // Used to map our severity from Phylum to valid Sarif
+    const validLevels = {
+        "critical": "error",
+        "high": "warning",
+        "medium": "note",
+        "low": "none",
+    };
+
+    const ruleIds: string[] = [];
+
+    for (const dep of input.dependencies) {
+        for (const rej of dep.issues) {
+            const ruleId = rej.tag;
+
+            // Add rule to the tool's rules if not already added
+            if (!ruleIds.includes(ruleId)) {
+                ruleIds.push(ruleId);
+                sarif.runs[0].tool.driver.rules.push({
+                    "id": ruleId,
+                    "name": rej.title,
+                    "shortDescription": {
+                        "text": rej.description
+                    },
+                    "defaultConfiguration": {
+                        "level": validLevels[rej.impact]
+                    }
+                });
+            }
+
+            // Add result
+            sarif.runs[0].results.push({
+                "ruleId": ruleId,
+                "message": {
+                    "text": rej.title
+                },
+                "locations": [
+                    {
+                        "physicalLocation": {
+                            "artifactLocation": {
+                                "uri": dep.purl
+                            }
+                        }
+                    }
+                ]
+            });
+        }
+    }
+
+    return sarif;
+};
+
+/** 
+ * Fetch the project ID for the provided project name;
+ */
+async function fetchProjectNames(group?: string): Promise<any> {
+    const url = group ? `groups/${group}/projects`: `data/projects/`;
```

Comment:
Is the trailing slash necessary? It seems odd to have it only on one side of the ternary operator.


```suggestion
    const url = group ? `groups/${group}/projects`: `data/projects`;
```